kensu/pyspark/spark_connector.py

```diff
@@ -0,0 +1,839 @@
+
+    # schema
+    fields = [
+        FieldDef(name=f.name(), field_type=f.fieldType(), nullable=f.nullable())
+        for f in schema.fields()
+    ]
+    py_schema = Schema(
+        name=schema.name(),
+        pk=SchemaPK(py_ds.to_ref(),
+                    fields=fields))
+
+    def f_publish_stats(lineage_run_id):
+        schema.attachStatsTo(lineage_run_id)
+
+    def f_get_stats():
+        # ******************************************************************************************
+        # WARNING : Do not return anything else than None, because f_publish_stats is preferred giving the responsibility
+        # to Spark to send the data stats only when it's available (Future computation is completed)
+        # ******************************************************************************************
+        return None
+    return KensuDatasourceAndSchema(ksu_ds=py_ds, ksu_schema=py_schema, f_get_stats=f_get_stats, f_publish_stats = f_publish_stats)
+
+
+def j2py_dict_of_lists(jdict):
+    r = {}
+    for k, jlist in jdict.items():
+        r[k] = list(jlist)
+    return r
+
+
+def get_inputs_lineage_fn(dam, df):
+    jvm = get_jvm_from_df(df)
+    # call def fetchToPandasReport(
+    #     df: DataFrame,
+    #     name: String = UUID.randomUUID().toString,
+    #     logicalName: Option[String] = None,
+    #     datasourceType: String = InternalDatasourceTags.TAG_INPUT_LINEAGE_ONLY,
+    #     timeout: Duration
+    #   )
+    dam_client_class = ref_scala_object(jvm, "io.kensu.third.integration.dataconversions.SparkToPandasConvTracker")
+    duration_300s =  ref_scala_object(jvm, "scala.concurrent.duration.Duration").apply(300, "second")
+    # FIXME: it should NOT wait for datasts -> check
+    import uuid
+    virtual_ds_name = str(uuid.uuid1())
+    no_logical_name = jvm.scala.Option.apply(None)
+    lineage_from_scala = dam_client_class.fetchToPandasReport(
+        df._jdf,
+        virtual_ds_name,
+        no_logical_name,
+        # ds_type,
+        duration_300s
+    )
+    # logging.info('lineage_from_scala:', lineage_from_scala)
+    df.lineage_from_scala = lineage_from_scala
+    from kensu.utils.dsl.extractors.external_lineage_dtos import GenericComputedInMemDs, ExtDependencyEntry
+    # case class SimpleInputDsSchemaAndLineage(
+    #   ds: DataSource,
+    #   schema: Schema,
+    #   lineage: Map[String, Seq[String]]
+    # )
+    #
+    # case class SimpleInputsLineageReport(
+    #   lineage: Seq[SimpleInputDsSchemaAndLineage]
+    # )
+    lineage_info = [
+        ExtDependencyEntry(
+            input_ds=scala_ds_and_schema_to_py(dam, e.ds(), e.schema()),
+            lineage=j2py_dict_of_lists(e.lineage())
+        ) for e in lineage_from_scala.lineage()
+    ]
+    logging.info('lineage_info:', lineage_info)
+    return GenericComputedInMemDs(inputs=list([x.input_ds for x in lineage_info]), lineage=lineage_info)
+
+
+def kwargs_or_conf_or_default(key, default, arg=None, kw=None, conf=None, tpe=None):
```

Comment:
isn't this a copy-paste from kensu.py (or whatever used when initialized kensu-py conf) ?